### ISSUE TITLE
Error in calculating correlations for a 6x6 spin system

### ISSUE THREAD

**Issue by adithi16** [weinbe58/QuSpin#448]
We are using quspin to solve a 2D 6x6 sites spin model with the following symmetries: Translational symmetries in x and y, Zero magnetisation sector, Spin inversion and Simultaneous parity along x and y directions. We are first finding the ground state in the symmetry reduced basis and proceeding to find the spin-spin spatial correlation function after projecting it back to the full basis(with zero magnetisation) using "project_from(pcon=True)".
The code gives expected results for smaller systems (like 4x4, 6x4, 6x5) but fails to give correct correlations for 6x6. This is the snippet of the code that calculates the correlations after finding the ground state psi_GS_red. 

```
basis_p = spin_basis_general(N, m=0, S='1/2', kxblock=(T_x,0),kyblock=(T_y,0),zblock=(Z,0),Pblock=(PxPy,0),pauli=False)
basis_full= spin_basis_general(N,m=0, S="1/2", pauli=False)
psi_GS_full= basis_p.project_from(psi_GS_red, sparse = False, pcon = True)
Sz_0_list = [["z", [0], 1]]
Sz0_GS = basis_full.inplace_Op(psi_GS_full, Sz_0_list, np.complex64)
Corr_full = np.zeros((Ly,Lx),dtype=np.float32)
for i in range(N):
    print(i)
    Sz_r_list = [["z", [i], 1]]
    Szr_GS = basis_full.inplace_Op(psi_GS_full, Sz_r_list, np.complex64)
    Corr_full[y[i], x[i]] = 3*np.dot(Szr_GS.conj(),Sz0_GS)
```
We also tried getting the expectation value of SzSz at a given site in the full basis using both quantum_LinearOperator and hamiltonian. Both these work fine for smaller system size but shows error for 6x6. 
This is the part of the code that gives the expectation value.

```
Sz= [[1, 0, 0]]
static1= [["zz", Sz]]
dynamic1= []
Sz0Sz0 = hamiltonian(static1,dynamic1, basis=basis_full, dtype=np.complex64)
exp= Sz0Sz0.expt_value(psi_GS_full)

```

For 6x6, these are errors for quantum_LinearOperator and hamitonian: 
```
Sz0Sz0 = quantum_LinearOperator(static1, basis=basis_full, dtype=np.complex64)
Particle conservation check passed!
Traceback (most recent call last):
  File "Correlation_calc_expt.py", line 75, in <module>
    Sz0Sz0 = quantum_LinearOperator(static1, basis=basis_full, dtype=np.complex64)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/operators/quantum_LinearOperator_core.py", line 133, in _init_
    ME,row,col = self.basis.Op(opstr,indx,J,self._dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/basis/base.py", line 218, in Op
    return self._Op(opstr,indx,J,dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/basis/basis_general/spin.py", line 181, in _Op
    ME,row,col = hcb_basis_general._Op(self,opstr,indx,J,dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/basis/basis_general/base_general.py", line 274, in _Op
    self._core.op(row,col,ME,opstr,indx,J,self._basis,self._n,self._basis_begin,self._basis_end,self._N_p)
  File "quspin/basis/basis_general/basis_general_core/source/general_basis_core.pyx", line 224, in quspin.basis.basis_general._basis_general_core.hcb_core.general_basis_core_wrap._pyx_fused_cpdef
TypeError: Function call with ambiguous argument types
```

```
Sz0Sz0 = hamiltonian(static1,dynamic1, basis=basis_full, dtype=np.complex64)
Particle conservation check passed!
Traceback (most recent call last):
  File "Correlation_calc_expt.py", line 76, in <module>
    Sz0Sz0 = hamiltonian(static1,dynamic1, basis=basis_full, dtype=np.complex64)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/operators/hamiltonian_core.py", line 322, in _init_
    self._static=make_static(self._basis,static_opstr_list,dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/operators/_make_hamiltonian.py", line 109, in make_static
    return basis._make_matrix(static_list,dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/basis/base.py", line 226, in _make_matrix
    ME,row,col = self.Op(opstr,indx,J,dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/basis/base.py", line 218, in Op
    return self._Op(opstr,indx,J,dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/basis/basis_general/spin.py", line 181, in _Op
    ME,row,col = hcb_basis_general._Op(self,opstr,indx,J,dtype)
  File "/home/samudrasur/anaconda3/lib/python3.7/site-packages/quspin/basis/basis_general/base_general.py", line 274, in _Op
    self._core.op(row,col,ME,opstr,indx,J,self._basis,self._n,self._basis_begin,self._basis_end,self._N_p)
  File "quspin/basis/basis_general/basis_general_core/source/general_basis_core.pyx", line 224, in quspin.basis.basis_general._basis_general_core.hcb_core.general_basis_core_wrap._pyx_fused_cpdef
TypeError: Function call with ambiguous argument types
```
These error files are also attached. 
P.S. We cannot use op_shift sector method since non-zero momentum and parity do not commute.   

![Corr_error](https://user-images.githubusercontent.com/79824047/132046856-1b5465a4-14a9-4a9c-a7c8-b714220039af.png)

**Comment by mgbukov** [comment]
@adithi16 it's hard to help you, unless you provide us with more information. In particular, we need

0) the Hamiltonian of your system
1) the data file with the stored ground state for a system of 4x4 and 6x6: if it is too large, please provide a link to a data repo to download it from
2) python scripts that load the ground state and compute the correlation function: you can upload them here (have to change the file extension from .py to .txt)
3) the correct value for the correlation function for 4x4 and 6x6 (so we can compare against it), and a desciption of the procedure how it was determined.

**Comment by adithi16** [comment]
0 . Our system is a 2D triangular lattice of spin half particles with different Heisenberg exchange couplings depending on the orientation of the three bonds. 

1 . & 2. Attached is tar.gz file containing two folders with codes and final correlators. (We are unable to upload the ground state file for 6x6 due to its large size of about 500MB) 
The code "ground_state_calc.py" computes the ground state for given parameters and stores it in "GS_4x4.npy". The code "Correlation_calc.py" reads this file and calculates the two spin correlations for all the sites. This correlation matrix is stored in "Corr_4x4". 
Similar codes and files have been attached for the 6x6 system.  

3 . (a) The correlation for S_{0}S_{0} (on site 0) should always give 0.75 (3/4) irrespective of the system size and parameters. As you can see in "Corr_4x4", for 4x4 this is gives correct result. But the first element of "Corr_6x6" is not 3/4. 
We do get 3/4 for all other smaller system size configurations. 
(b) Also, for equal couplings, like in the codes attached, the correlation functions between adjacent spins to site 0 should be equal in all three directions. This is also seen for 4x4 but it is not the case for 6x6. 

[quspin_doubt.tar.gz](https://github.com/weinbe58/QuSpin/files/7110195/quspin_doubt.tar.gz)

**Comment by mgbukov** [comment]
thanks, I'll take a look! 

Would it be possible for you to provide a link to the file with GS data for 6x6, which I can download (e.g. using Google drive, etc.)?

And three questions: 
i) why do u use dype=complex64, if the Hamiltonian is real-valued? 
ii) why do you work with single rather than double precision, i.e. why np.complex64 vs np.complex128?
iii) what is the value that your correlator gives at 6x6 instead of 3/4?

**Comment by adithi16** [comment]
This is the link to the ground state file. Please let us know if you can access it. 

https://drive.google.com/file/d/1kS7p7ktm3CUWGY3_EQMYl-0CiDFVSRth/view?usp=sharinghttps://drive.google.com/file/d/1kS7p7ktm3CUWGY3_EQMYl-0CiDFVSRth/view?usp=sharing 

(i) (ii) Using either float64, complex64 or complex128 gives the same result. We checked with all dtypes. 
(iii)1.180228739976882935e-01 is the value we are getting instead of 3/4. This is the first element in the "Corr_6x6" file. This value is 0.75 in "Corr_4x4" às you can see.

**Comment by mgbukov** [comment]
@adithi16 thanks, I got the GS!

@weinbe58 this looks like a problem with `inplace_Op` when the system size exceeds 30 sites; can it be that there is nome problem with 64 bit integers there (6x6 requires 64 bits)?

**Comment by adithi16** [comment]
@mgbukov @weinbe58 Is there a way to fix this 64-bit problem?

**Comment by mgbukov** [comment]
it is not yet clear to me that the problem is with the 64-bit integers in `inplace_Op`: if you do the same calculation in another magnetization sector at 6x6 where the H-space is much smaller, the result is correct. So the culprit might also be `project_from` (also in conjunction with the 64-bit integers).

I'm trying to reproduce the issue on the cluster, but I am not sure there is a large enough memory node there to compute the full basis.

**Comment by mgbukov** [comment]
@adithi16 I have problems reproducing this bug; I ran your code with your GS data on the cluster, and I obtained the correct result for the equal-site correlation function.

Here's the output I get (I am using the basis with Pauli matrices instead of spin-1/2 matrices, and I removed the factor of 3 in the last line of the correlator, so that the Corr(0,0)=1.0):

```
63092837
9075135300
norm psi_GS_red: 0.9999999999999989
norm psi_GS_full: 0.9999999973075766
correlator 0 1.0
```

The remaining values for the sigma_z sigma_z correlator are stored in this file (change the file extension back to .npy):

[results_6x6.txt](https://github.com/weinbe58/QuSpin/files/7120658/results_6x6.txt)

Moreover, since the value for Corr(0,0) is independent of the state the expectation is computed in, I also tried a completely random state, which produced the same value:

```
random: norm psi_GS_red: 0.9999999999999992
random: norm psi_GS_full: 0.9999999999702476
0 (0, 0) 1.0000000000000013
```

Here's my script:

[Correlation_calc_py.txt](https://github.com/weinbe58/QuSpin/files/7120688/Correlation_calc_py.txt)

**Comment by yonasG** [comment]
@mgbukov or @adithi16, I am using the example provided above to learn how to calculate the static-spin structure factor (SSSF). Assuming my physics is correct, I have some questions about the how to extend the spin correlation results described above. I have attached a pdf describing my goal and my attempts.  I have also attached the code in two versions. Any help you can give me is appreciated.

[Questions.pdf](https://github.com/weinbe58/QuSpin/files/7230339/Questions.pdf)

[SSSF_try.txt](https://github.com/weinbe58/QuSpin/files/7230340/SSSF_try.txt)

**Comment by adithi16** [comment]
@mgbukov Thank you very much for your reply. Your code works fine on our system too. We were unable to figure out what was wrong with our code! 
However since this method also took a lot of time of our cluster, and we wanted to produce correlations for a large set of parameters, we came up with an alternate method. 
We stored the projectors connecting two basis using `get_proj`. 
P1 with *Basis_1* with translation, spin inversion, zero magnetisation and reflection which doesnot commute for any momentum hence calculating correlations was a problem. 
P2 *Basis_2* with everything **except** reflection in which we could calculate the correlations using `op_shift'. 
After calculating the eigenstates along with reflection symmetry *Basis_1*, we projected it to a *Basis_2* using P2^{dagger}P1
which was stored and loaded everytime. This was way faster and also gave accurate results. 
However, thank you for your help and discussion. 
We are very sorry we could not get back for so long due to some issues.
